stackslib/src/net/httpcore.rs

```diff
@@ -0,0 +1,1625 @@
+                self.last_four_preamble_bytes[3],
+                buf[0],
+            ],
+            2 => [
+                self.last_four_preamble_bytes[2],
+                self.last_four_preamble_bytes[3],
+                buf[0],
+                buf[1],
+            ],
+            3 => [self.last_four_preamble_bytes[3], buf[0], buf[1], buf[2]],
+            _ => [buf[i - 4], buf[i - 3], buf[i - 2], buf[i - 1]],
+        };
+        window
+    }
+
+    /// Given a fully-formed single HTTP response, parse it (used by clients).
+    #[cfg(test)]
+    pub fn parse_response(
+        verb: &str,
+        request_path: &str,
+        response_buf: &[u8],
+    ) -> Result<StacksHttpMessage, NetError> {
+        let mut http = StacksHttp::new(
+            "127.0.0.1:20443".parse().unwrap(),
+            &ConnectionOptions::default(),
+        );
+
+        let response_handler_index =
+            http.find_response_handler(verb, request_path)
+                .ok_or(NetError::SendError(format!(
+                    "No such handler for '{} {}'",
+                    verb, request_path
+                )))?;
+        http.request_handler_index = Some(response_handler_index);
+
+        let (preamble, message_offset) = http.read_preamble(response_buf)?;
+        let is_chunked = match preamble {
+            StacksHttpPreamble::Response(ref resp) => resp.is_chunked(),
+            _ => {
+                return Err(NetError::DeserializeError(
+                    "Invalid HTTP message: did not get a Response preamble".to_string(),
+                ));
+            }
+        };
+
+        let mut message_bytes = &response_buf[message_offset..];
+
+        if is_chunked {
+            match http.stream_payload(&preamble, &mut message_bytes) {
+                Ok((Some((message, _)), _)) => Ok(message),
+                Ok((None, _)) => Err(NetError::UnderflowError(
+                    "Not enough bytes to form a streamed HTTP response".to_string(),
+                )),
+                Err(e) => Err(e),
+            }
+        } else {
+            let (message, _) = http.read_payload(&preamble, &mut message_bytes)?;
+            Ok(message)
+        }
+    }
+}
+
+impl ProtocolFamily for StacksHttp {
+    type Preamble = StacksHttpPreamble;
+    type Message = StacksHttpMessage;
+
+    /// how big can a preamble get?
+    fn preamble_size_hint(&mut self) -> usize {
+        HTTP_PREAMBLE_MAX_ENCODED_SIZE as usize
+    }
+
+    /// how big is this message?  Might not know if we're dealing with chunked encoding.
+    fn payload_len(&mut self, preamble: &StacksHttpPreamble) -> Option<usize> {
+        match *preamble {
+            StacksHttpPreamble::Request(ref http_request_preamble) => {
+                Some(http_request_preamble.get_content_length() as usize)
+            }
+            StacksHttpPreamble::Response(ref http_response_preamble) => {
+                match http_response_preamble.content_length {
```

Comment:
Can remove this match and just do a map:
`http_response_preamble.content_length.map(|len| len as usize)`

Reply:
Fixed